### PR TITLE
TFAC-177/178/219/226 Create the UserImpactTable and mutations to modify the fields

### PR DESCRIPTION
**Pull request by amaliwanag2**

### CODE REVIEW

File: dynamodb/tables.json [hunk 1]
```diff
@@ -38,6 +38,21 @@
       "WriteCapacityUnits": 1
     }
   },
+  {
+    "TableName": "UserImpact",
+    "KeySchema": [
+      { "AttributeName": "id", "KeyType": "HASH" },
```
Comment (kmjennison) [inline dynamodb/tables.json:44]: @amaliwanag2 This attribute should be `userId`.

File: graphql/database/userImpact/UserImpactModel.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        .required()
+        .description(`the charity id that maps to the charity table`),
+      userImpactMetric: types
+        .number()
+        .default(self.fieldDefaults.userImpactMetric)
+        .description(
+          `the metric a user contributes to a charity.  For example, it might be number of treats for
+          cats.`
+        ),
+      pendingUserReferralImpact: types
```
Comment (kmjennison) [inline graphql/database/userImpact/UserImpactModel.js:47]: Do we also need `pendingUserReferrals` to know the number of referrals to display?

File: graphql/database/userImpact/updateImpact.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import UserImpactModel from './UserImpactModel'
+import { USER_VISIT_IMPACT_VALUE } from '../constants'
+/**
+ * updates a user impact record
+ * @param {object} userContext - The user authorizer object.
+ * @param {string} userId - The user id.
+ * @param {string} tabId - A UUID for this opened tab
```
Comment (kmjennison) [inline graphql/database/userImpact/updateImpact.js:7]: Fix "tabId" here

File: graphql/database/userImpact/UserImpactModel.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      visitsUntilNextImpact: types
+        .number()
+        .default(self.fieldDefaults.visitsUntilNextImpact) // only set in app code
+        .description(`A counter that decrements each time a user tabs until it reaches 0 and
+        the user impact is incremented`),
+      confirmedImpact: types
+        .boolean()
+        .default(self.fieldDefaults.confirmedImpact)
+        .description(`enables a user to start earning user impact once they have opted in
+        on the UI`),
```
Comment (kmjennison) [inline graphql/database/userImpact/UserImpactModel.js:61]: And do we need a flag to know if the frontend should show a "congrats, you reached X impact"?

File: graphql/database/userImpact/updateImpact.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ * @param {object} updates - options to update impact record, include logImpact, claimPendingReferralImpact, confirmImpact
+ * @return {Promise<User>}  A promise that resolves into a UserImpact instance.
+ */
+
+const updateImpact = async (userContext, userId, charityId, updates) => {
```
Comment (kmjennison) [inline graphql/database/userImpact/updateImpact.js:12]: I see what you're saying about this one mutation doing many things :) It's fine to leave for now, though I agree it's easier to reason about mutations and schema when a mutation is responsible for only one action.

File: graphql/database/userImpact/__tests__/updateImpact.test.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+import updateImpact from '../updateImpact'
+import { USER_VISIT_IMPACT_VALUE } from '../../constants'
+import {
+  DatabaseOperation,
+  getMockUserContext,
+  mockDate,
+  setMockDBResponse,
+} from '../../test-utils'
+// We use an export from the mock file.
+// eslint-disable-next-line import/named
```
Comment (kmjennison) [inline graphql/database/userImpact/__tests__/updateImpact.test.js:13]: Unneeded eslint and comment?

File: graphql/database/userImpact/UserImpactModel.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    }
+  }
+
+  static get permissions() {
+    return {
+      get: permissionAuthorizers.userIdMatchesHashKey,
+      getAll: () => false,
+      update: permissionAuthorizers.userIdMatchesHashKey,
+      // To create a new user, the created item must have the same
+      // email and user ID as the authorized user.
```
Comment (kmjennison) [inline graphql/database/userImpact/UserImpactModel.js:80]: Remove unrelated comment

**Review by kmjennison — CHANGES_REQUESTED** [review]
Changes are only for a few stray comments/docs. Left a couple of questions just to raise whether there's additional fields we need, but I'll defer to your planning on this, and you can merge this before adding those (if needed).

File: graphql/database/userImpact/UserImpactModel.js (hunk 4 above)
Comment (amaliwanag2) [reply graphql/database/userImpact/UserImpactModel.js:61]: yup!

File: graphql/database/userImpact/UserImpactModel.js (hunk 2 above)
Comment (amaliwanag2) [reply graphql/database/userImpact/UserImpactModel.js:47]: yeah I was assuming we'd deduce that based on impact but it makes sense to store both

File: dynamodb/tables.json [hunk 8]
```diff
@@ -38,6 +38,21 @@
       "WriteCapacityUnits": 1
     }
   },
+  {
+    "TableName": "UserImpact",
```
Comment (kmjennison) [inline dynamodb/tables.json:42]: Need this same info in serverless.yml.

**Review by kmjennison — CHANGES_REQUESTED** [review]
Excellent work. Only change: please add the new table to serverless.yml as well so we can get it live in the cloud. You can set a very low minimum read/write capacity (probably just 1 for each), given the scaling will be slow.

File: graphql/serverless.yml [hunk 9]
```diff
@@ -153,6 +153,17 @@ provider:
       Resource:
         - "${self:custom.dbTableArnBase}BackgroundImages-${self:provider.stage}"
         - "${self:custom.dbTableArnBase}BackgroundImages-${self:provider.stage}/index/*"
+    # UserImpact table
```
Comment (amaliwanag2) [inline graphql/serverless.yml:156]: @kmjennison ah, thanks for catching

**Review by kmjennison — APPROVED** [review]

**Pull request merged**